File: Make Sum Divisible by P/Make Sum Divisible by P.py
# Solution for Make Sum Divisible by P in PY

class Solution:
    def minSubarray(self, nums, p):
        n = len(nums)

        # total sum mod p
        total_mod_sum = sum(nums) % p

        if total_mod_sum == 0:
            return 0

        required_remainder = total_mod_sum % p

        prefix_mod_to_index = {0: -1}
        current_prefix_mod = 0
        min_length = n

        for j, num in enumerate(nums):
            current_prefix_mod = (current_prefix_mod + num) % p
            needed_mod = (current_prefix_mod - required_remainder) % p

            if needed_mod in prefix_mod_to_index:
                min_length = min(min_length, j - prefix_mod_to_index[needed_mod])

            prefix_mod_to_index[current_prefix_mod] = j

        return -1 if min_length == n else min_length

File: Make Sum Divisible by P/Make Sum Divisible by P.java
// Solution for Make Sum Divisible by P in JAVA

class Solution {
    public int minSubarray(int[] nums, int p) {
        int n = nums.length;

        // Compute total sum mod p
        int totalModSum = 0;
        for (int num : nums) {
            totalModSum = (totalModSum + num) % p;
        }

        if (totalModSum == 0) return 0;

        int requiredRemainder = totalModSum % p;

        // Map prefix_mod → index
        Map<Integer, Integer> prefixModToIndex = new HashMap<>();
        prefixModToIndex.put(0, -1);

        int currentPrefixMod = 0;
        int minLength = n;

        for (int j = 0; j < n; j++) {
            currentPrefixMod = (currentPrefixMod + nums[j]) % p;
            int neededMod = (currentPrefixMod - requiredRemainder + p) % p;

            if (prefixModToIndex.containsKey(neededMod)) {
                minLength = Math.min(minLength, j - prefixModToIndex.get(neededMod));
            }

            prefixModToIndex.put(currentPrefixMod, j);
        }

        return minLength == n ? -1 : minLength;
    }
}
